server/routerlicious/packages/services/src/mongoExceptionRetryRules/mongoError/index.ts

```diff
@@ -0,0 +1,172 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IMongoExceptionRetryRule } from "../IMongoExceptionRetryRule";
+class InternalErrorRule implements IMongoExceptionRetryRule {
+    private static readonly codeName = "InternalError";
+    match(error: any): boolean {
+        return error.code === 1
+            && error.codeName
+            && (error.codeName as string) === InternalErrorRule.codeName;
+    }
+
+    shouldRetry: boolean = false;
+}
+
+// This handles the requested queued on client side buffer overflow. Should relies on reconnect instead of retry?
+class NoConnectionAvailableRule implements IMongoExceptionRetryRule {
+    private static readonly messagePrefix = "no connection available for operation and number of stored operation";
+    match(error: any): boolean {
+        // TODO: This timed out actually included two different messages:
+        // 1. Retries due to rate limiting: False.
+        // 2. Retries due to rate limiting: True.
+        // We might need to split this into two different rules after consult with DB team.
+        return error.message
+            && (error.message as string).startsWith(NoConnectionAvailableRule.messagePrefix);
+    }
+
+    shouldRetry: boolean = false;
+}
+
+// This handles the no primary found in replicaset or invalid replica set name from client
+// Should not retry but relays on reconnect.
+class NoPrimaryInReplicasetRule implements IMongoExceptionRetryRule {
+    private static readonly message = "no primary found in replicaset or invalid replica set name";
+    match(error: any): boolean {
+        // TODO: This timed out actually included two different messages:
+        // 1. Retries due to rate limiting: False.
+        // 2. Retries due to rate limiting: True.
+        // We might need to split this into two different rules after consult with DB team.
+        return error.message
+            && (error.message as string) === NoPrimaryInReplicasetRule.message;
+    }
+
+    shouldRetry: boolean = false;
+}
+
+// this handles the pool destroyed error from client side. Should relies on reconnect instead of retry?
```

Comment:
Are these comments final, or are we still investigating the answer for these questions?

Reply:
These are errors from client side driver, for now I have no clue